runtime/src/mips.rs

```diff
@@ -0,0 +1,721 @@
+        /// corresponds ot the dispatchable action and vote with some balance.
+        ///
+        /// # Arguments
+        /// * `proposal` a dispatchable call
+        /// * `index` proposal index
+        /// * `aye_or_nay` a bool representing for or against vote
+        /// * `deposit` minimum deposit value
+        #[weight = SimpleDispatchInfo::FixedNormal(200_000)]
+        pub fn vote(origin, proposal_hash: T::Hash, index: MipsIndex, aye_or_nay: bool, deposit: BalanceOf<T>) {
+            let proposer = ensure_signed(origin)?;
+
+            let mut voting = Self::voting(&proposal_hash).ok_or("proposal does not exist")?;
+            ensure!(voting.index == index, "mismatched proposal index");
+
+            let position_yes = voting.ayes.iter().position(|(a, _)| a == &proposer);
+            let position_no = voting.nays.iter().position(|(a, _)| a == &proposer);
+
+            if position_yes.is_none() && position_no.is_none()  {
+                if aye_or_nay {
+                    voting.ayes.push((proposer.clone(), deposit));
+                } else {
+                    voting.nays.push((proposer.clone(), deposit));
+                }
+                <Voting<T>>::remove(&proposal_hash);
+                <Voting<T>>::insert(&proposal_hash, voting);
+                Self::deposit_event(RawEvent::Voted(proposer, proposal_hash, aye_or_nay));
+            } else {
+                return Err("duplicate vote ignored")
+            }
+        }
+
+        /// An emergency stop measure to kill a proposal. Governance committee can kill
+        /// a proposal at any time.
+        #[weight = SimpleDispatchInfo::FixedNormal(200_000)]
+        pub fn kill_proposal(origin, proposal_hash: T::Hash) {
+            T::CommitteeOrigin::try_origin(origin)
+                .map(|_| ())
+                .or_else(ensure_root)
+                .map_err(|_| "bad origin")?;
+
+            Self::close_proposal(proposal_hash.clone());
+        }
+
+        /// Moves a referendum instance into dispatch queue.
+        #[weight = SimpleDispatchInfo::FixedNormal(200_000)]
+        pub fn enact_referundum(origin, proposal_hash: T::Hash) {
+            T::CommitteeOrigin::try_origin(origin)
+                .map(|_| ())
+                .or_else(ensure_root)
+                .map_err(|_| "bad origin")?;
+
+            Self::prepare_to_dispatch(proposal_hash);
+        }
+
+        /// When constructing a block check if it's time for a ballot to end. If ballot ends,
+        /// proceed to ratification process.
+        fn on_initialize(n: T::BlockNumber) {
+            if let Err(e) = Self::end_block(n) {
+                sr_primitives::print(e);
+            }
+        }
+
+    }
+}
+
+impl<T: Trait> Module<T> {
+    /// Retrieve all proposals that need to be closed as of block `n`.
+    pub fn proposals_maturing_at(n: T::BlockNumber) -> Vec<(MipsIndex, T::Hash)> {
+        Self::proposal_meta()
```

Comment:
This seems very expensive to run at the end of each block (potentially, depending on the number of outstanding proposals). Would it be better to have a mapping from end_block_number => Vec<Proposals> so that we can skip directly to the proposals which end on a particular block?

Reply:
I've thought about `end_block_number => Vec` mapping. That makes sense in scenarios where we know for sure proposal needs to be called at a certain block *and* we don't have to cancel a proposal. It's all or nothing with no way to back out. Since we let proposals to be cancelled, this is a compromise that allows lookup and yet iterate over the proposals.